### ISSUE TITLE
Add support for Regex script

### ISSUE THREAD

**Issue by arthur-fontaine** [pnpm/pnpm#7081]
### Contribution

- [X] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

As a user of an "external script runner", it can be very verbose to create a new script.

For example, if you are using [Wireit](https://github.com/google/wireit), you first have to add a key-value to the `scripts` object in the `package.json`, then you can write your script. Same problem with Turbo.

### Describe the solution you'd like

In the same way that pnpm allows Regex to be used in the `run` command, we could add support for "Regex as a key" in the `scripts` object in the package.json.

For example:

```json
{
  ...,
  "scripts": {
    "/.*/": "echo \"I run for every command\""
  }
}
```

Running `pnpm run a` or `pnpm run b` or any other script will always echo "I run for every command".

> [!NOTE]
> If there is both a Regex script and a literal script, both scripts will run.

---

We could also add support for substitution.

For example:

```json
{
  ...,
  "scripts": {
    "/.*/": "turbo run \\0"
  }
}
```

Running `pnpm run a` will run `turbo run a`

> [!IMPORTANT]
> I use "\\digit" to match the pattern as `sed` does. There may be a better pattern to use.

### Describe the drawbacks of your solution

pnpm already allows the use of Regex in the `run` command (https://pnpm.io/cli/run#running-multiple-scripts). These features may overlap.

If "Regex in scripts" is implemented, I think that using a Regex in the `run` command should match the source of the regex.

Example:

```json
{
  ...,
  "scripts": {
    "/.*/": "echo \"I run for every command\"",
    "test:a": "echo \"test a\"",
    "test:b": "echo \"test b\""
  }
}
```

_Result:_
| Command | Result |
| --- | --- |
| `pnpm run "/.*/"` | *run all scripts* |
| `pnpm run "/test:*/"` | *run `test:a` and `test:b`* |
| `pnpm run "/\/\.\*\//"` | *run `/.*/`* |

### Describe alternatives you've considered

I have found no alternative to this feature.

**Comment by KSXGitHub** [comment]
There is a point where it is better to use a proper taskrunner (such as a `make` or `just`) than to rely on the package manager to do everything that isn't related to managing packages.

Anyway, I don't those should be in `scripts`. Maybe `regexScripts` or `pnpm.regexScripts`?

**Comment by arthur-fontaine** [comment]
Why do you think it should not be in `scripts`?
`regexScripts` is not in the [package.json specs](https://docs.npmjs.com/cli/v10/configuring-npm/package-json) and in my opinion we should avoid breaking them as much as possible.
Using `scripts` will not break the package.json specs. pnpm will just behave differently than npm. I think that this feature could also be enabled through the user config to prevent breaks even more.

**Comment by KSXGitHub** [comment]
> Why do you think it should not be in `scripts`?

Because `scripts` aren't supposed to be regex.

> `regexScripts` is not in the [package.json specs](https://docs.npmjs.com/cli/v10/configuring-npm/package-json) and in my opinion we should avoid breaking them as much as possible.

There's also `pnpm.regexScripts`.

Anyway, your use-case is to redirect all `pnpm run` to another taskrunner, right? How about `pnpm.taskRunner` which would be invoked if `pnpm run` does not find matching script name in `scripts`?

**Comment by zkochan** [comment]
I agree that this is out of scope for pnpm. We concentrate on package management not running tasks. If you need such complex tasks configuration, maybe you should use a dedicated tool or create your own and use it instead of `pnpm run`.

**Comment by everett1992** [comment]
This doesn't seem compatible with pnpm's regex to run multiple tasks - if the regex is '/test:.*/' and turbo has test:a, test:b, test:c pnpm could not enumerate a, b, c. 

why aren't you calling turbo directly, why do you need to enter thru pn and package.json scripts?
